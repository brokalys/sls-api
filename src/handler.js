import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
