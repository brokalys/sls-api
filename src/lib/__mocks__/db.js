export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
