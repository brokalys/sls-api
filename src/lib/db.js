import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
