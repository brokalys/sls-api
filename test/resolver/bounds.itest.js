import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';

const { query } = createTestClient(server);

describe('Resolver: bounds', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  test('successfully retrieves bounds data', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBounds($bounds: String!) {
          bounds(bounds: $bounds) {
            bounds
          }
        }
      `,
      variables: {
        bounds:
          '56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves all the possible data', async () => {
    authenticateAs('slsCrawler', server);

    const response = await query({
      query: `
        query ($bounds: String!) {
          bounds(bounds: $bounds) {
            bounds
            buildings {
              id
              bounds
              properties {
                results {
                  id
                  url
                  category
                  type
                  rent_type
                  content
                  images
                  price
                  calc_price_per_sqm
                  rooms
                  area
                  floor
                  lat
                  lng
                  published_at
                  foreign_id
                }
                summary {
                  count
                  price {
                    count
                    min
                    max
                    mean
                    median
                    mode
                    standardDev
                  }
                }
              }
            }
            land {
              id
              bounds
              properties {
                results {
                  id
                  url
                  category
                  type
                  rent_type
                  content
                  images
                  price
                  calc_price_per_sqm
                  rooms
                  area
                  floor
                  lat
                  lng
                  published_at
                  foreign_id
                }
                summary {
                  count
                  price {
                    count
                    min
                    max
                    mean
                    median
                    mode
                    standardDev
                  }
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('fails retrieving all the data due to missing permissions', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query ($bounds: String!) {
          bounds(bounds: $bounds) {
            bounds
            buildings {
              id
              bounds
              properties {
                results {
                  id
                  url
                  category
                  type
                  rent_type
                  content
                  images
                  price
                  calc_price_per_sqm
                  rooms
                  area
                  floor
                  lat
                  lng
                  published_at
                  foreign_id
                }
                summary {
                  count
                  price {
                    count
                    min
                    max
                    mean
                    median
                    mode
                    standardDev
                  }
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully returns nothing if there are no buildings in this bound', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBuildings($bounds: String!) {
          bounds(bounds: $bounds) {
            buildings {
              id
              properties {
                results {
                  price
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves building and property information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBuildingsAndProperties($bounds: String!) {
          bounds(bounds: $bounds) {
            buildings {
              id
              properties {
                results {
                  price
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves land and property information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query($bounds: String!) {
          bounds(bounds: $bounds) {
            land {
              id
              properties {
                results {
                  price
                }
              }
              vzd {
                land {
                  sale_id
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if no bounds filter provided', async () => {
    const response = await query({
      query: `
        {
          bounds {
            bounds
          }
        }
      `,
    });
    authenticateAs('mapApp', server);

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if bounds filter has too large location', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBounds($bounds: String!) {
          bounds(bounds: $bounds) {
            bounds
            buildings{
              properties {
                results {
                  id
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if bounds filter has too large location, but allows for larger region if no property data requested', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBounds($bounds: String!) {
          bounds(bounds: $bounds) {
            bounds
          }
        }
      `,
      variables: {
        bounds:
          '57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await query({
      query: `
        query GetBuildingsAndProperties($bounds: String!) {
          bounds(bounds: $bounds) {
            buildings {
              id
              properties {
                results {
                  price
                }
              }
            }
          }
        }
      `,
      variables: {
        bounds:
          '56.944756215513316 24.09404948877113, 56.9404253703127 24.09404948877113, 56.9404253703127 24.086952363717725, 56.944756215513316 24.086952363717725, 56.944756215513316 24.09404948877113',
      },
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });
});
