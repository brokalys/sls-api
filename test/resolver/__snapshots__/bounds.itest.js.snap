// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver: bounds fails retrieving all the data due to missing permissions 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "message": "You need following role: ReadPropertyData:Detailed",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds successfully retrieves all the possible data 1`] = `
Object {
  "data": Object {
    "bounds": Object {
      "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619",
      "buildings": Array [
        Object {
          "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
          "id": 1,
          "properties": Object {
            "results": Array [
              Object {
                "area": 90,
                "calc_price_per_sqm": 1.11,
                "category": "apartment",
                "content": "",
                "floor": 2,
                "foreign_id": null,
                "id": 1,
                "images": Array [],
                "lat": 56.1,
                "lng": 24.9,
                "price": 100,
                "published_at": "2021-04-12T19:47:30.000Z",
                "rent_type": "monthly",
                "rooms": null,
                "type": "rent",
                "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
              },
              Object {
                "area": null,
                "calc_price_per_sqm": null,
                "category": "apartment",
                "content": "",
                "floor": null,
                "foreign_id": null,
                "id": 2,
                "images": Array [],
                "lat": 56.1,
                "lng": 24.9,
                "price": 200,
                "published_at": "2021-04-12T19:47:30.000Z",
                "rent_type": "monthly",
                "rooms": null,
                "type": "rent",
                "url": "https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html",
              },
            ],
            "summary": Object {
              "count": 2,
              "price": Object {
                "count": 2,
                "max": 200,
                "mean": 150,
                "median": 150,
                "min": 100,
                "mode": 100,
                "standardDev": 50,
              },
            },
          },
        },
        Object {
          "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
          "id": 2,
          "properties": Object {
            "results": Array [
              Object {
                "area": null,
                "calc_price_per_sqm": null,
                "category": "house",
                "content": "",
                "floor": null,
                "foreign_id": null,
                "id": 3,
                "images": Array [],
                "lat": null,
                "lng": null,
                "price": 300,
                "published_at": "2021-04-12T19:47:30.000Z",
                "rent_type": "monthly",
                "rooms": null,
                "type": "rent",
                "url": "https://www.ss.com/msg/lv/real-estate/homes-summer-residences/tukums-and-reg/engures-pag/hhjbf.html",
              },
            ],
            "summary": Object {
              "count": 1,
              "price": Object {
                "count": 1,
                "max": 300,
                "mean": 300,
                "median": 300,
                "min": 300,
                "mode": 300,
                "standardDev": null,
              },
            },
          },
        },
        Object {
          "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
          "id": 3,
          "properties": Object {
            "results": Array [],
            "summary": Object {
              "count": 0,
              "price": Object {
                "count": 0,
                "max": null,
                "mean": null,
                "median": null,
                "min": null,
                "mode": null,
                "standardDev": null,
              },
            },
          },
        },
      ],
      "land": Array [
        Object {
          "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
          "id": 1,
          "properties": Object {
            "results": Array [
              Object {
                "area": 90,
                "calc_price_per_sqm": 1.11,
                "category": "apartment",
                "content": "",
                "floor": 2,
                "foreign_id": null,
                "id": 1,
                "images": Array [],
                "lat": 56.1,
                "lng": 24.9,
                "price": 100,
                "published_at": "2021-04-12T19:47:30.000Z",
                "rent_type": "monthly",
                "rooms": null,
                "type": "rent",
                "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
              },
              Object {
                "area": null,
                "calc_price_per_sqm": null,
                "category": "apartment",
                "content": "",
                "floor": null,
                "foreign_id": null,
                "id": 2,
                "images": Array [],
                "lat": 56.1,
                "lng": 24.9,
                "price": 200,
                "published_at": "2021-04-12T19:47:30.000Z",
                "rent_type": "monthly",
                "rooms": null,
                "type": "rent",
                "url": "https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html",
              },
            ],
            "summary": Object {
              "count": 2,
              "price": Object {
                "count": 2,
                "max": 200,
                "mean": 150,
                "median": 150,
                "min": 100,
                "mode": 100,
                "standardDev": 50,
              },
            },
          },
        },
        Object {
          "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
          "id": 2,
          "properties": Object {
            "results": Array [],
            "summary": Object {
              "count": 0,
              "price": Object {
                "count": 0,
                "max": null,
                "mean": null,
                "median": null,
                "min": null,
                "mode": null,
                "standardDev": null,
              },
            },
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds successfully retrieves bounds data 1`] = `
Object {
  "data": Object {
    "bounds": Object {
      "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619, 56.992294 24.136619",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds successfully retrieves building and property information 1`] = `
Object {
  "data": Object {
    "bounds": Object {
      "buildings": Array [
        Object {
          "id": 1,
          "properties": Object {
            "results": Array [
              Object {
                "price": 100,
              },
              Object {
                "price": 200,
              },
            ],
          },
        },
        Object {
          "id": 2,
          "properties": Object {
            "results": Array [
              Object {
                "price": 300,
              },
            ],
          },
        },
        Object {
          "id": 3,
          "properties": Object {
            "results": Array [],
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds successfully retrieves land and property information 1`] = `
Object {
  "data": Object {
    "bounds": Object {
      "land": Array [
        Object {
          "id": 1,
          "properties": Object {
            "results": Array [
              Object {
                "price": 100,
              },
              Object {
                "price": 200,
              },
            ],
          },
          "vzd": Object {
            "land": Array [
              Object {
                "sale_id": 1,
              },
            ],
          },
        },
        Object {
          "id": 2,
          "properties": Object {
            "results": Array [],
          },
          "vzd": Object {
            "land": Array [
              Object {
                "sale_id": 1,
              },
              Object {
                "sale_id": 100,
              },
            ],
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds successfully returns nothing if there are no buildings in this bound 1`] = `
Object {
  "data": Object {
    "bounds": Object {
      "buildings": Array [],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds throws a validation exception if bounds filter has too large location 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "area": 0.0001,
              "key": "bounds",
              "label": "bounds",
              "value": "57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768",
            },
            "message": "\\"bounds\\" are must not exceed 0.0001",
            "path": Array [
              "bounds",
            ],
            "type": "string.maxArea",
          },
        ],
      },
      "message": "Input validation failed",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds throws a validation exception if bounds filter has too large location, but allows for larger region if no property data requested 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "area": 0.001,
              "key": "bounds",
              "label": "bounds",
              "value": "57.0510741522279 24.34369621296768, 56.86735048784755 24.34369621296768, 56.86735048784755 23.842917061051175, 57.0510741522279 23.842917061051175, 57.0510741522279 24.34369621296768",
            },
            "message": "\\"bounds\\" are must not exceed 0.001",
            "path": Array [
              "bounds",
            ],
            "type": "string.maxArea",
          },
        ],
      },
      "message": "Input validation failed",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: bounds throws a validation exception if no bounds filter provided 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "message": "Field \\"bounds\\" argument \\"bounds\\" of type \\"String!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
